src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +19,53 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHourUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(currCurrent_>=0)
+    {
+        return false;
+    }
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double ampHoursLeft;
+    int current = currCurrent_;
+    if(isCharging())
+    {
+        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal_;
+        current *= -1;
+    }
+    else
+    {
+        ampHoursLeft = ampHourTotal_;
+    }
+    double msecLeft = (ampHoursLeft * HOURS_TO_MSECS) / current;
+    QTime QTimeLeft = QTime(0,0,0,0).addMSecs(msecLeft);
+    return QTimeLeft;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if(currTime_.isNull())
+    {
+        prevTime_ = QTime(0,0,0,0);
+    }
+    prevCurrent_ = currCurrent_;
+    currCurrent_ = batteryData.current;
+    prevTime_ = QTime(currTime_.hour(),currTime_.minute(),currTime_.second(),currTime_.msec());
```

Comment:
You can just use the copy constructor here, `prevTime = currTime;` would suffice.